csrc/ir/nodes.cpp

```diff
@@ -2114,16 +2114,19 @@ std::vector<PolymorphicValue> MmaOp::evaluate(
   auto output = in_a.matmul(in_b);
 
   // ATen preserves the input dtype whereas MmaOP generates float outputs.
-  // Cast to the dtype of the MmaOp output for consistency.
-  // NOTE: MmaOp returns the float output, whereas in the evaluate method,
-  //      we are casting from float -> input_dtype -> float. This will lead
-  //      to loss of precision.
-  //      MmaOp::evaluate should be modified to effectively handle cast(MmaOp(H,
-  //      H), H) This will avoid the above cast chain and precision issue.
-  if (tv_a->getDataType() != out()->getDataType().value()) {
-    output = output.to(data_type_to_aten(out()->getDataType().value()));
-  }
-  return {output};
+  // If the MmaOp is followed by a cast, skip casting back the input to avoid
+  // roundtrip casts. For eg: MmaOp (H, H) + CastOp (F, H) will result in
+  // F->H->F->H impacting precision and performance. Note: This potentially
+  // binds the MmaOp output in the expression evaluator to a tensor
+  //        of different datatype.
+  auto uses = out()->uses();
+  if ((tv_a->getDataType() == out()->getDataType().value()) ||
+      (uses.size() == 1 && uses.front()->isA<UnaryOp>() &&
+       uses.front()->as<UnaryOp>()->getUnaryOpType() == UnaryOpType::Cast)) {
```

Comment:
Is it necessary to also check that the single use casts to `tv_a->getDataType()`?

Reply:
I don't think it is. We can let the cast handle the final dtype.
The cases where we won't cast back to float are `half, bfloat16` for matmul. So even if the single use was casting to float instead of the tv_a dtype, it would work.

Reply:
For example, `BFloat(MmaOp(Half,Half))` is not equivalent to `Bfloat(Half(MmaOp(Half,Half)))`, what `at::matmul` + `at::Tensor::to` would compute. Because `Half` has more mantissa bits and `Bfloat` has more exponent bits, the extra conversion to `Half` is going to lose precision or even worse clip large values to `inf`. 

What options do we have to precisely expression-evaluate `BFloat(MmaOp(Half,Half))`? Making it `mul+sum+cast` is the only option I can think of. If we don't want to include that option in this particular PR, I'd still make this function hard-fail on things like `BFloat(MmaOp(Half,Half))`. Because it's much easier to root-cause a hard-fail than a numeric mismatch.

Reply:
If you agree with above, I'd simplify the code to

```
NVF_CHECK(out() is eventually used as output.dtype());
return output;
```

Reply:
I considered that case, but didn't think it occured in practice. OOC, are there models/cases where we see this?
I would cast back to float in this case instead of erroring out, wdyt? So we skip the cast if and only if it is a single cast to the same dtype.

Reply:
> OOC, are there models/cases where we see this?

No. FYI, for cases that haven't occurred, assuming similar code complexity, my general preference is
1. to generate functional and performant code 
2. to generate functional but slow code
3. to hard fail during compilation (earlier is better)
4. to generate code that hard fails (earlier is better)
5. to generate code that produces wrong numbers

> I would cast back to float in this case instead of erroring out, wdyt? 

Can you clarify? 

Do you mean `BFloat(MmaOp(Half,Half))` becomes `at::matmul + to(float)`? Wouldn't that cause the fusion to output a different dtype than torch? 

Or, do you mean `BFloat(MmaOp(Half,Half))` becomes `at::matmul + to(float) + to(bfloat)`? That IIUC would have the same precision loss problem as `Bfloat(Half(MmaOp(Half,Half)))`, because of the cast to `half` (done by `at::matmul`).

Reply:
> Or, do you mean `BFloat(MmaOp(Half,Half))` becomes `at::matmul + to(float) + to(bfloat)`? That IIUC would have the same precision loss problem as `Bfloat(Half(MmaOp(Half,Half)))`, because of the cast to `half` (done by `at::matmul`).

You're right, we would run into the precision issue here, although we are no less precise than pytorch in this case. The cast to `float` is however unnecessary.

> What options do we have to precisely expression-evaluate BFloat(MmaOp(Half,Half))? Making it mul+sum+cast is the only option I can think of. 

Will this be done in the matmul scheduler?

>If we don't want to include that option in this particular PR, I'd still make this function hard-fail on things like BFloat(MmaOp(Half,Half)). Because it's much easier to root-cause a hard-fail than a numeric mismatch.

Will error out in this PR.

Reply:
> If you agree with above, I'd simplify the code to
> 
> ```
> NVF_CHECK(out() is eventually used as output.dtype());
> return output;
> ```

What do you mean? If the cast is to the correct input dtype, then the output.dtype() is different than out().

Reply:
By "eventually", I meant the dtype after the cast. For example, in `Half(MmaOp(H,H))`, MmaOp's out is used as a half after the cast, so it's safe to simply return output. (Of course, the single-use check is still necessary)

Reply:
> Will this be done in the matmul scheduler?

No, I was thinking of doing it in MmaOp::evaluate. Actually, we might not need to decompose. Can you confirm `MmaOp(H,H)` is numerically equivalent to `at::matmul(Float(H),Float(H))`? 

(Since we are converging to let such cases error out, the above question is not so relevant any more)

Reply:
> By "eventually", I meant the dtype after the cast. For example, in `Half(MmaOp(H,H))`, MmaOp's out is used as a half after the cast, so it's safe to simply return output. (Of course, the single-use check is still necessary)

We should also allow cast to `float`. In the current setting, if we see a cast to double, that will also error out. Is this a case we we want to support? 
IMO cast to double following MmaOp will be unlikely in practice so I am fine with not increasing the list of conditions on the cast output type. Wdyt?

Reply:
> We should also allow cast to float

I don't think so. Casting to float or double should error out as well for the same reason. `Double(MmaOp(H,H))` is not equivalent to `at::matmul(H,H).double()`. `Float(MmaOp(H,H))` is not equivalent to `at::matmul(H,H).float()`.

Reply:
I think the confusion comes from that we use different definitions for "precision". Sorry, I should have caught that earlier. I'm not talking about difference being less than a pre-defined threshold. I'm talking about numeric equivalence where `a+b == b+a` but `float(half(a)) != a`. 

Numeric equivalence is important, because I saw and debugged too many cases where subtle numeric inequivalence between two implementations gets amplified by many layers of a network, leading to vastly different end results and hurting model accuracy. 

That being said, numeric equivalent alone doesn't guarantee bit exactness. Other things like accumulation order can be different between implementations, leading to different bitwise results. For example, a CPU implementation of a matmul likely adopts a different partition strategy than a GPU one. However, numeric equivalence rules out many other sources of differences and often makes the difference between implementations manageable (e.g. the auto threshold thing you worked on in testValidate). 

Now back to the practical stuff -- what we should do with the existing tests (SingleMmaOp and MatmulWithBias). I'd change SingleMmaOp to test mma+cast instead (and therefore probably change the test name as well). I'd disable MatmulWithBias until we have at::addmm hooked up with ExpressionEvaluator. PS: to disable the test, you can just change the name to `DISABLED_MatmulWithBias` and googletest does the magic to skip it and show it in the disabled test list.

Reply:
Got it, thanks for the explanation. I see your point now.